### PR TITLE
[MachineInstr] add insert method for variadic instructions

### PR DESCRIPTION
**Pull request by nickdesaulniers**
As alluded to in #20571, it would be nice if we could mutate operand
lists of MachineInstr's more safely. Add an insert method that together
with removeOperand allows for easier splicing of operands.

Splitting this patch off early to get feedback; I need to either:
- mutate an INLINEASM{_BR} MachinInstr's MachineOperands from being
  registers (physical or virtual) to memory
  (MachineOperandType::MO_FrameIndex).  These are not 1:1 operand
  replacements, but N:M operand replacements. i.e. we need to
  update 2 MachineOperands into the middle of the operand list to 5 (at
  least for x86_64).
- copy, modify, write a new MachineInstr which has its relevant operands
  replaced.

Either approaches are hazarded by existing references to either the
operands being moved, or the instruction being removed+replaced. For my
purposes in regalloc, either seem to work for me, so hopefully reviewers
can help me determine which approach is preferable. The second would
involve no new methods on MachineInstr.

One question I had while looking at this was: "why does MachineInstr
have BOTH a NumOperands member AND a MCInstrDesc member that itself has
a NumOperands member? How many operands can a MachineInstr have? Do I
need to update BOTH (keeping them in sync)?" FWICT, only "variadic"
MachineInstrs have MCInstrDesc with NumOperands (of the MCInstrDesc) set
to zero.  If the MCInstrDesc's NumOperands is non-zero, then the NumOperands
on the MachineInstr itself cannot exceed this value (IIUC) else an assert will
be triggered.

For most non-psuedo instructions (or at least non-varidic instructions),
insert is less likely to be useful.

To run the newly added unittest:
    $ pushd llvm/build; ninja CodeGenTests; popd
    $ ./llvm/build/unittests/CodeGen/CodeGenTests \
        --gtest_filter=MachineInstrTest.SpliceOperands

This is meant to mirror `MCInst::insert`.

### CODE REVIEW

**Comment by nickdesaulniers** [comment]
Added to the description:

> This is meant to mirror MCInst::insert.

Also, should I limit the new ops to not be "implicit" ops? Looking at `MachineInstr::addOperand`, it seems like where they are located is not in a specific ordering relative to the "explicit" ops.

File: llvm/lib/CodeGen/MachineInstr.cpp [hunk 1]
```diff
@@ -2475,3 +2475,24 @@ MachineInstr::getFirst5RegLLTs() const {
...
+  const unsigned OpsToMove = NumOperands - OpIdx;
+
+  SmallVector<MachineOperand> MovingOps;
+  MovingOps.reserve(OpsToMove);
+
+  for (unsigned I = 0; I < OpsToMove; ++I) {
+    MovingOps.emplace_back(getOperand(OpIdx));
+    removeOperand(OpIdx);
+  }
+  for (const MachineOperand &MO : Ops)
```
Comment (nickdesaulniers) [inline llvm/lib/CodeGen/MachineInstr.cpp:2494]: TODO: should we assert that each MO is not implicit?

**Comment by qcolombet** [comment]
> One question I had while looking at this was: "why does MachineInstr
> have BOTH a NumOperands member AND a MCInstrDesc member that itself has
> a NumOperands member? 

The short answer is:
`MCInstrDesc` holds what is known statically.
`NumOperands` holds how many actual operands there are attached to this particular `MachineInstr` at compile time.

E.g., consider a `PHI` instruction.
Its `MCInstrDesc` is going to describe the number of operands we **must** have and that's `1` (the definition of the `PHI`).
Now the `NumOperands` of the related instantiation of a `PHI` will report the definition and all the arguments.

Generally speaking, there is a guarantee that `MI`'s `NumOperands` >= `MCInstrDesc`'s `NumOperands`.
(See https://github.com/llvm/llvm-project/blob/32d16b64d3125e76f65d7d88a302a33618eb0e6e/llvm/lib/CodeGen/MachineVerifier.cpp#L1828)

> How many operands can a MachineInstr have?

Unlimited :).

> Do I
> need to update BOTH (keeping them in sync)?"

`MCInstrDesc`'s `NumOperands` cannot be changed, so if you need to keep something in sync that means you need to set a different `MCInstrDesc` on the related `MI`. (E.g., go from `ADDI32rr` to `ADDI32rm`.)

`MI`'s `NumOperands` is updated automatically through the calls to `removeOperand` and `addOperand`.

> FWICT, only "variadic"
> MachineInstrs have MCInstrDesc with NumOperands (of the MCInstrDesc) set
> to zero. If the MCInstrDesc's NumOperands is non-zero, then the NumOperands
> on the MachineInstr itself cannot exceed this value (IIUC) else an assert will
> be triggered.

Hmm that doesn't sound right.
Where do you see this assert?

IIRC the only thing we check is that the number of definitions and the number of arguments are at least as big as what `MCInstrDesc` requires.
Then, of course, there's a bunch of other checks around register classes and whatnot, but with respect to the number of operands, that's about it I think.

Long story short, for inline asm, I don't think you have to worry about `NumOperands`.

File: llvm/lib/CodeGen/MachineInstr.cpp [hunk 2]
```diff
@@ -2475,3 +2475,24 @@ MachineInstr::getFirst5RegLLTs() const {
       Reg2, getRegInfo()->getType(Reg2), Reg3, getRegInfo()->getType(Reg3),
       Reg4, getRegInfo()->getType(Reg4));
 }
+
+void MachineInstr::insert(mop_iterator It, ArrayRef<MachineOperand> Ops) {
+  assert(isVariadic() && "can only modify variadic instructions");
```
Comment (MatzeB) [inline llvm/lib/CodeGen/MachineInstr.cpp:2480]: Couldn't we just as well add more implicit operands to a non-variadic instrution or similar? Could be implicit operands or maybe there is a situation where not all operands have been added to an instruction yet. So if possible I'd just remove this constraint.

File: llvm/lib/CodeGen/MachineInstr.cpp [hunk 3]
```diff
@@ -2475,3 +2475,24 @@ MachineInstr::getFirst5RegLLTs() const {
...
+  assert(It->getParent() == this && "iterator points to operand of other inst");
```
Comment (MatzeB) [inline llvm/lib/CodeGen/MachineInstr.cpp:2481]: Probably want to check this for every operand in the loop below?

File: llvm/lib/CodeGen/MachineInstr.cpp (hunk 1 above)
Comment (MatzeB) [reply llvm/lib/CodeGen/MachineInstr.cpp:2494]: I don't see why you would check. I'd just treat it as a list of operands. The fact that you need to respect some ordering etc. I would leave to callers of the APIs (or verifiers to catch problems).

**Review by MatzeB — COMMENTED** [review]
Adding an API to insert operands in the middle of the operand list seems fine. I'd just leave semantics like implicit, variadic etc. out of it (and make that a problem for callers and verifiers).

**Comment by nickdesaulniers** [comment]
Marking WIP as I've found one hazard with this interface.

Tied operands don't like being removed.  We need to untie them, then re-tie them around the operands list updates.

File: llvm/lib/CodeGen/MachineInstr.cpp (hunk 2 above)
Comment (nickdesaulniers) [reply llvm/lib/CodeGen/MachineInstr.cpp]: done in https://github.com/llvm/llvm-project/pull/67699/commits/92d53e9c3cefdb7433a3caff4a681f722d70298d

File: llvm/lib/CodeGen/MachineInstr.cpp (hunk 3 above)
Comment (nickdesaulniers) [reply llvm/lib/CodeGen/MachineInstr.cpp]: I don't follow; we need to check that the iterator for where in a given MachineInstr's list of operands to insert into points into that MachineInstr's operand list and not some other random/unrelated MachineIntr's operand list.

**Comment by nickdesaulniers** [comment]
Sorry for the delays. Was "wrastling" with register allocators this past week.  This is ready for review.

File: llvm/lib/CodeGen/MachineInstr.cpp (hunk 3 above)
Comment (MatzeB) [reply llvm/lib/CodeGen/MachineInstr.cpp]: I think I was thinking of something like this to make sure operands don't belong to some other instruction yet:
```
  for (MachineOperand* MOP : Ops) { assert(MOP->getParen() == nullptr); }
```
(which is a slightly different check from the iterator one you had there; sorry for confusion)

Comment (MatzeB) [reply llvm/lib/CodeGen/MachineInstr.cpp]: Then again, the code ends up calling `addOperand()` and such an assert should rather be within that function then. So nothing to do here I guess.

Comment (MatzeB) [reply llvm/lib/CodeGen/MachineInstr.cpp]: Also scratch this whole comment. It seems addOperand() and this insert() here actually copies the operands by value so no need for such a check at all.

File: llvm/lib/CodeGen/MachineInstr.cpp [hunk 4]
```diff
@@ -2475,3 +2475,45 @@ MachineInstr::getFirst5RegLLTs() const {
...
+  const unsigned OpIdx = getOperandNo(It);
+  const unsigned NumOperands = getNumOperands();
+  const unsigned OpsToMove = NumOperands - OpIdx;
+
+  SmallVector<MachineOperand> MovingOps;
+  MovingOps.reserve(OpsToMove);
+
+  for (unsigned I = 0; I < OpsToMove; ++I) {
+    MovingOps.emplace_back(getOperand(OpIdx));
+    removeOperand(OpIdx);
```
Comment (MatzeB) [inline llvm/lib/CodeGen/MachineInstr.cpp:2504]: Maybe do this in reverse order so we waste less cycles on the `moveOperands()` call within `MachineInstr::removeOperand()`.   On that note I wonder if `moveOperands()` would work here as well instead of the `removeOperand` / `addOperand` for `MovingOps` here?

Admittedly it looks like `moveOperands()` doesn't really adjust the `TiedTo` markers so that's why it wouldn't work here?

**Comment by nickdesaulniers** [comment]
Spooky, looks like my newly added tests are failing in presubmit.

When I run them locally via:
```
$ ./llvm/build/unittests/CodeGen/CodeGenTests --gtest_filter=MachineInstrTest.SpliceOperands
```
I observe no issues.  When I run them via:
```
$ ./llvm/build/unittests/CodeGen/CodeGenTests
```
sometimes I observe a segfault, sometimes I observe
```
CodeGenTests: /android0/llvm-project/llvm/include/llvm/CodeGen/MachineOperand.h:370: Register llvm::MachineOperand::getReg() const: Assertion `isReg() && "This is not a register operand!"' failed.
```
Is there some kind of race on MachineRegisterInfo for these tests?

File: llvm/lib/CodeGen/MachineInstr.cpp [hunk 5]
```diff
@@ -2475,3 +2475,45 @@ MachineInstr::getFirst5RegLLTs() const {
       Reg2, getRegInfo()->getType(Reg2), Reg3, getRegInfo()->getType(Reg3),
       Reg4, getRegInfo()->getType(Reg4));
 }
+
+void MachineInstr::insert(mop_iterator It, ArrayRef<MachineOperand> Ops) {
+  if (!It || Ops.empty())
+    return;
+
+  // Do one pass to untie operands.
+  DenseMap<unsigned, unsigned> TiedOpIndices;
```
[the remark this answers is not in the document]
Comment (nickdesaulniers) [reply llvm/lib/CodeGen/MachineInstr.cpp:2484]: https://github.com/llvm/llvm-project/pull/70677

File: llvm/lib/CodeGen/MachineInstr.cpp (hunk 4 above)
Comment (nickdesaulniers) [reply llvm/lib/CodeGen/MachineInstr.cpp:2504]: > On that note I wonder if moveOperands() would work here as well instead of the removeOperand / addOperand for MovingOps here?

The issue there is we would miss the calls to `MachineFunction::allocateOperandArray` from `MachineInstr::addOperand`. Since splice grows the operand list, we need to realloc the storage for the operands.  `MachineInstr::moveOperands` just `memmove`s operands around which is not safe to do when the backing store hasn't been grown.

> Admittedly it looks like moveOperands() doesn't really adjust the TiedTo markers so that's why it wouldn't work here?

That's also an issue.

> Maybe do this in reverse order so we waste less cycles on the moveOperands() call within MachineInstr::removeOperand().

I don't think there are any cycles to be saved? Perhaps if splice is being used to append operands on the end?  Usually you splice operands into the middle, so you need to memmove the elements occurring after the insertion point multiple times.

Notice in this loop that the induction variable `I` is not being used as an index; we're removing `OpIdx` repeatedly as this is slicing off the elements past the insertion point.

Later, we need to re-add the operands back in the correct order (though it's not hard to reverse the loop below).

Did I perhaps miss something thought?

File: llvm/lib/CodeGen/MachineInstr.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (nickdesaulniers) [reply llvm/lib/CodeGen/MachineInstr.cpp]: done in https://github.com/llvm/llvm-project/pull/67699/commits/6ec41786ae1a0a1e169e48cf84080a9ed8b16979

File: llvm/lib/CodeGen/MachineInstr.cpp [hunk 6]
```diff
@@ -2475,3 +2475,45 @@ MachineInstr::getFirst5RegLLTs() const {
...
+      unsigned TiedTo = findTiedOperandIdx(OpNo);
+      TiedOpIndices[OpNo] = TiedTo;
+      untieRegOperand(OpNo);
+    }
+  }
+
+  assert(It->getParent() == this && "iterator points to operand of other inst");
+  const unsigned OpIdx = getOperandNo(It);
+  const unsigned NumOperands = getNumOperands();
+  const unsigned OpsToMove = NumOperands - OpIdx;
```
[the remark this answers is not in the document]
Comment (nickdesaulniers) [reply llvm/lib/CodeGen/MachineInstr.cpp]: done in https://github.com/llvm/llvm-project/pull/67699/commits/381dd4c3e0eae7510e48611aa13c01f9cc435fb8

File: llvm/lib/CodeGen/MachineInstr.cpp [hunk 7]
```diff
@@ -2475,3 +2475,45 @@ MachineInstr::getFirst5RegLLTs() const {
...
+  for (const MachineOperand &MO : operands()) {
+    if (MO.isReg() && MO.isTied()) {
+      unsigned OpNo = getOperandNo(&MO);
+      unsigned TiedTo = findTiedOperandIdx(OpNo);
+      TiedOpIndices[OpNo] = TiedTo;
+      untieRegOperand(OpNo);
+    }
+  }
+
+  assert(It->getParent() == this && "iterator points to operand of other inst");
```
[the remark this answers is not in the document]
Comment (nickdesaulniers) [reply llvm/lib/CodeGen/MachineInstr.cpp]: done in https://github.com/llvm/llvm-project/pull/67699/commits/69b94e345e87ec19af8898053ae9e4b8da208b63

File: llvm/lib/CodeGen/MachineInstr.cpp [hunk 8]
```diff
@@ -2475,3 +2475,45 @@ MachineInstr::getFirst5RegLLTs() const {
       Reg2, getRegInfo()->getType(Reg2), Reg3, getRegInfo()->getType(Reg3),
       Reg4, getRegInfo()->getType(Reg4));
 }
+
+void MachineInstr::insert(mop_iterator It, ArrayRef<MachineOperand> Ops) {
+  if (!It || Ops.empty())
```
[the remark this answers is not in the document]
Comment (nickdesaulniers) [reply llvm/lib/CodeGen/MachineInstr.cpp]: done in https://github.com/llvm/llvm-project/pull/67699/commits/4a862dd7a3c7e544ed13836a7f94f2ccf2df154c

File: llvm/include/llvm/CodeGen/MachineInstr.h [hunk 9]
```diff
@@ -1818,6 +1818,9 @@ class MachineInstr
   /// preferred.
   void addOperand(const MachineOperand &Op);
 
+  /// Inserts Ops BEFORE It. Can untie/retie tied operands.
+  void insert(mop_iterator It, ArrayRef<MachineOperand> Ops);
```
Comment (MatzeB) [inline llvm/include/llvm/CodeGen/MachineInstr.h:1822]: Another compiler I worked on would always call these sort of parameters `InsertBefore` or `InsertAfter` (instead of `It` here) to make the semantics clear. Maybe that would be neat here to show the semantics? On the other hand I haven't seen this pattern anywhere else in LLVM so no strong opinion.

File: llvm/lib/CodeGen/MachineInstr.cpp (hunk 4 above)
Comment (MatzeB) [reply llvm/lib/CodeGen/MachineInstr.cpp:2505]: make sense. I haven't looked at the code as deeply as you have; just wanted to ask the questions to make sure :)

**Review by MatzeB — APPROVED** [review]
LGTM

File: llvm/include/llvm/CodeGen/MachineInstr.h (hunk 9 above)
Comment (nickdesaulniers) [reply llvm/include/llvm/CodeGen/MachineInstr.h:1822]: yeah, I was thinking about that this weekend; would be a nice change. Will do.

**Comment by nickdesaulniers** [comment]
> Spooky, looks like my newly added tests are failing in presubmit.

Still a draft because of this. I wonder if my non-`MachineInstrBuilder` raw usage of the `MachineInstr` is problematic.

FWICT, these two lines will sometimes crash (sometimes depends which):
```c++
MI->getOperand(0).ChangeToRegister(1, /*isDef=*/true);                                                                                     
MI->getOperand(1).ChangeToRegister(2, /*isDef=*/false);
```
I wonder if `-DLLVM_ENABLE_ASSERTIONS=ON` isn't being applied to the unittests?

**Comment by MatzeB** [comment]
I'd be extremely surprised if we mess up the assertion setting for unit tests.

Have you tried adding a call to `verifyMachineFunction()` yet? Sometimes that uncovers more problems...

**Comment by nickdesaulniers** [comment]
> Spooky, looks like my newly added tests are failing in presubmit.

The call to `getRegUseDefListHead` in `MachineRegisterInfo::addRegOperandToUseList` is returning nullptr.  I must have to inform MRI of the new reg BEFORE adding it as an operand.

**Comment by nickdesaulniers** [comment]
> Spooky, looks like my newly added tests are failing in presubmit.
> The call to getRegUseDefListHead in MachineRegisterInfo::addRegOperandToUseList is returning nullptr. I must have to inform MRI of the new reg BEFORE adding it as an operand. I'll add an assert for that.

aha! It's that a tied operands are generally a def (output) followed by a use (input); if the input/use is added without a corresponding def, MRI has conniptions.

File: llvm/include/llvm/CodeGen/MachineInstr.h (hunk 9 above)
Comment (nickdesaulniers) [reply llvm/include/llvm/CodeGen/MachineInstr.h]: done in https://github.com/llvm/llvm-project/pull/67699/commits/6ba0e06260532ca0cd0e2b54b0f8968951d2be23

**Comment by nickdesaulniers** [comment]
> Is there some kind of race on MachineRegisterInfo for these tests?

fixed in https://github.com/llvm/llvm-project/pull/67699/commits/84496264cb40938476a22f840b8a449844893dbe, will wait for a green presubmit run before submitting.
